RFC.md

```diff
@@ -0,0 +1,413 @@
+
+class FullModelTorchTuneCheckpointer(_CheckpointerInterface):
+	...
+
+    def load_checkpoint(self, weights_only: bool = True) -> Dict[str, Any]:
+        ...
+
+		if "layers.0.sa_norm.scale" in model_state_dict:
+			logger.info(
+				"This is an older checkpoint format. Converting \
+				to the current version."
+			)
+			model_state_dict = convert_weights._legacy_to_tune(
+				model_state_dict, "0.2.1"
+			)
+			state_dict[utils.MODEL_KEY] = model_state_dict
+		...
+
+```
+## Caveats
+
+### Alternatives
+@kartikayk expressed concerns over exposing SelfAttention and CrossAttention at the build level which might add complexity for users to parse. While we still requires users to pass in the attention module anyway, another way we could approach this would be to parametrize TransformerDecoderLayer to handle both cross attention and self attention.
+
+```python
+class TransformerDecoderLayer(nn.Module):
+    def forward(
+        self,
+        x: Tensor,
+        *,
+        mask: Optional[Tensor] = None,
+        encoder_input: Optional[Tensor] = None,
+        encoder_mask: Optional[Tensor] = None,
+        input_pos: Optional[Tensor] = None,
+    ) -> Tensor:
+
+        if encoder_input is None and self.cross_attn_layer:
+            return x
+
+		# return None if not self.cross_attn_layer
+        skip_mask = self._skip_mask(encoder_mask)
+
+		y = encoder_input if self.cross_attn_layer else x
+		m = encoder_mask if self.cross_attn_layer else mask
+        attn_out = self.attn(
+            self.attn_norm(x),
+            y,
+            mask=m,
+            input_pos=input_pos
+        )
+        if skip_mask is not None:
+            attn_out.masked_fill_(skip_mask, 0)
+
+        # Residual connection; shape: [batch_size, seq_length, embed_dim]
+        h = self.attn_scale(attn_out) + x
+
+        # Norm applied before the feedforward layer
+        mlp_out = self.mlp(self.mlp_norm(h))
+        if skip_mask is not None:
+            mlp_out.masked_fill_(skip_mask, 0)
+
+        out = h + self.mlp_scale(mlp_out)
+        return out
+```
+
+This is a reasonable alternative but I do fear it'll get more and more complex to extend as we go forward and want to support more attention types.
+
+### Risks/Concerns
+- Attention module and model builders both get a bit more complex with this. The more variation we support the more complexity we introduce. I believe that supporting multiple transformer layers at least allows us to keep individual layers simpler.
+- This change breaks BC both in the checkpoint itself but also in model builders. We can update all of our model builders, but if users have their own custom ones they will have to update them too. Do we have a mechanism for clearly communicating this in the release notes along with how to update a model builder?
+- The cross attention layer here does not work with sample packing. Should we make layers that support sample packing through an error if they're used with packed data? @RdoubleA
```

Comment:
So to be clear, no sample packing will be enable for models that utilize cross attention?

Reply:
Not until sample packing is updated to update cross attention masks.